CI: gcc version changed to 13 on BLAS CI runs

I see gcc has been bumped from [gcc-11](https://github.com/numpy/numpy/actions/runs/11302482942/job/31438371668#step:5:22) to [gcc-13](https://github.com/numpy/numpy/actions/runs/11323991352/job/31487837097#step:5:22). This version change is causing various build problems since some default warnings changed into errors:

<details><summary>compile errors</summary>

```
../numpy/_core/src/multiarray/alloc.c: In function ‘PyDataMem_RENEW’:
../numpy/_core/src/multiarray/alloc.c:279:9: error: pointer ‘ptr’ may be used after ‘realloc’ [-Werror=use-after-free]
  279 |         PyTraceMalloc_Untrack(NPY_TRACE_DOMAIN, (npy_uintp)ptr);
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../numpy/_core/src/multiarray/alloc.c:277:14: note: call to ‘realloc’ here
  277 |     result = realloc(ptr, size);


../numpy/_core/src/umath/scalarmath.c.src: In function ‘byte_true_divide’:
../numpy/_core/src/umath/scalarmath.c.src:1389:23: error: ‘other_val_conv’ may be used uninitialized [-Werror=maybe-uninitialized]
 1389 |             other_val = other_val_conv;  /* Need a float value */
      |             ~~~~~~~~~~^~~~~~~~~~~~~~~~
../numpy/_core/src/umath/scalarmath.c.src:1357:14: note: ‘other_val_conv’ was declared here
 1357 |     @type@ other_val_conv;
      |              ^~~~~~~~~~~~~ 
../numpy/_core/src/umath/scalarmath.c.src: In function ‘short_true_divide’:
../numpy/_core/src/umath/scalarmath.c.src:1389:23: error: ‘other_val_conv’ may be used uninitialized [-Werror=maybe-uninitialized]
 1389 |             other_val = other_val_conv;  /* Need a float value */
      |             ~~~~~~~~~~^~~~~~~~~~~~~~~~
../numpy/_core/src/umath/scalarmath.c.src:1357:15: note: ‘other_val_conv’ was declared here
 1357 |     @type@ other_val_conv;
      |               ^~~~~~~~~~~~  
../numpy/_core/src/umath/scalarmath.c.src: In function ‘ushort_true_divide’:
../numpy/_core/src/umath/scalarmath.c.src:1389:23: error: ‘other_val_conv’ may be used uninitialized [-Werror=maybe-uninitialized]
 1389 |             other_val = other_val_conv;  /* Need a float value */
      |             ~~~~~~~~~~^~~~~~~~~~~~~~~~
../numpy/_core/src/umath/scalarmath.c.src:1357:16: note: ‘other_val_conv’ was declared here
 1357 |     @type@ other_val_conv;
      |                ^~~~~~~~~~~   
../numpy/_core/src/umath/scalarmath.c.src: In function ‘int_true_divide’:
../numpy/_core/src/umath/scalarmath.c.src:1389:23: error: ‘other_val_conv’ may be used uninitialized [-Werror=maybe-uninitialized]
 1389 |             other_val = other_val_conv;  /* Need a float value */
      |             ~~~~~~~~~~^~~~~~~~~~~~~~~~
../numpy/_core/src/umath/scalarmath.c.src:1357:13: note: ‘other_val_conv’ was declared here
 1357 |     @type@ other_val_conv;
      |             ^~~~~~~~~~~~~~
../numpy/_core/src/umath/scalarmath.c.src: In function ‘uint_true_divide’:
../numpy/_core/src/umath/scalarmath.c.src:1389:23: error: ‘other_val_conv’ may be used uninitialized [-Werror=maybe-uninitialized]
 1389 |             other_val = other_val_conv;  /* Need a float value */
      |             ~~~~~~~~~~^~~~~~~~~~~~~~~~
../numpy/_core/src/umath/scalarmath.c.src:1357:14: note: ‘other_val_conv’ was declared here
 1357 |     @type@ other_val_conv;
      |              ^~~~~~~~~~~~~ 
../numpy/_core/src/umath/scalarmath.c.src: In function ‘ulong_true_divide’:
../numpy/_core/src/umath/scalarmath.c.src:1389:23: error: ‘other_val_conv’ may be used uninitialized [-Werror=maybe-uninitialized]
 1389 |             other_val = other_val_conv;  /* Need a float value */
      |             ~~~~~~~~~~^~~~~~~~~~~~~~~~
../numpy/_core/src/umath/scalarmath.c.src:1357:15: note: ‘other_val_conv’ was declared here
 1357 |     @type@ other_val_conv;
      |               ^~~~~~~~~~~~  
../numpy/_core/src/umath/scalarmath.c.src: In function ‘ulonglong_true_divide’:
../numpy/_core/src/umath/scalarmath.c.src:1389:23: error: ‘other_val_conv’ may be used uninitialized [-Werror=maybe-uninitialized]
 1389 |             other_val = other_val_conv;  /* Need a float value */
      |             ~~~~~~~~~~^~~~~~~~~~~~~~~~
../numpy/_core/src/umath/scalarmath.c.src:1357:19: note: ‘other_val_conv’ was declared here
 1357 |     @type@ other_val_conv;
      |                   ^~~~~~~~      
```

</details>

--- a/numpy/_core/src/multiarray/alloc.c
+++ b/numpy/_core/src/multiarray/alloc.c
@@ -274,10 +274,8 @@ PyDataMem_RENEW(void *ptr, size_t size)
     void *result;
 
     assert(size != 0);
+    PyTraceMalloc_Untrack(NPY_TRACE_DOMAIN, (npy_uintp)ptr);
     result = realloc(ptr, size);
-    if (result != ptr) {
-        PyTraceMalloc_Untrack(NPY_TRACE_DOMAIN, (npy_uintp)ptr);
-    }
     PyTraceMalloc_Track(NPY_TRACE_DOMAIN, (npy_uintp)result, size);
     return result;
 }

--- a/numpy/_core/src/umath/scalarmath.c.src
+++ b/numpy/_core/src/umath/scalarmath.c.src
@@ -1354,7 +1354,7 @@ static PyObject *
      */
     PyObject *ret;
     npy_float64 arg1, arg2, other_val;
-    @type@ other_val_conv;
+    @type@ other_val_conv = 0;
 
     int is_forward;
     if (Py_TYPE(a) == &Py@Name@ArrType_Type) {

